Faster multiple file staging

lapfelix opened the pull request:
Staging/unstaging/discarding multiple files takes a long time. Seems to be O(N) where N is multiplied by 1-2 seconds, depending on the repo.

I fixed it by only calling `-[GCRepository writeRepositoryIndex:error:]` once when doing staging/discarding operations on multiple files by adding methods for these operations that take multiple files instead of one.

I tested the changes in a big repository (with thousands of commits, ~50 branches and ~20-30 total submodules) and the time it takes to stage/unstage 19 files went from ~25.5 seconds to ~2.5 seconds. The more files are modified, the bigger the performance increase is.

I AGREE TO THE GITUP CONTRIBUTOR LICENSE AGREEMENT

lucasderraugh commented:
Absolutely wonderful! I’m out for Christmas but I will gladly review this when I’m back on the 27th. 🍻

lucasderraugh commented on GitUpKit/Extensions/GCRepository+Index.m:
```diff
@@ -39,22 +39,61 @@ - (BOOL)resetIndexToHEAD:(NSError**)error {
+  for (NSString* path in paths) {
+    if (![self removeFile:path fromIndex:index error:error] || (error && *error != nil)) {
+      [self writeRepositoryIndex:index error:error];
+      return false;
```
`return NO`

lucasderraugh commented on GitUpKit/Extensions/GCRepository+Index.m:
```diff
@@ -39,22 +39,61 @@ - (BOOL)resetIndexToHEAD:(NSError**)error {
+    if (shouldWriteRepository) {
+      [self writeRepositoryIndex:index error:NULL];
+    }
+    return FALSE;
```
`return NO`

lapfelix replied:
ugh I started this branch 3 years ago and it shows 😂 I should've caught these when reviewing though

lucasderraugh commented on GitUpKit/Extensions/GCRepository+Index.m:
```diff
@@ -39,22 +39,61 @@ - (BOOL)resetIndexToHEAD:(NSError**)error {
+      break;
+    }
+
+    shouldWriteRepository = true;
```
`= YES`

lucasderraugh commented on GitUpKit/Utilities/GIViewController+Utilities.h:
```diff
@@ -48,11 +48,14 @@ extern NSString* const GIViewController_TerminalTool_iTerm;
+- (void)stageAllChangesForFiles:(NSArray<NSString*>*)paths;
 - (void)stageSelectedChangesForFile:(NSString*)path oldLines:(NSIndexSet*)oldLines newLines:(NSIndexSet*)newLines;
 - (void)unstageAllChangesForFile:(NSString*)path;
+- (void)unstageAllChangesForFiles:(NSArray<NSString*>*)path;
```
`paths`

lucasderraugh commented on GitUpKit/Views/GIAdvancedCommitViewController.m:
```diff
@@ -187,16 +187,24 @@ - (void)diffFilesViewControllerDidChangeSelection:(GIDiffFilesViewController*)co
+        [nonSubmoduleDeltas addObject:delta];
       }
-      [deltas addObject:delta];
     }
```
Seems like we can just have 2 arrays, `deltas` and `nonSubmoduleDeltasPaths`. In the end we're adding all the `deltas` together but we keep track of `nonSubmoduleDeltasPaths` to stage them all in bulk. I don't see why we need `nonSubmoduleDeltas`.

lapfelix replied:
Ah yes that totally makes sense

lucasderraugh commented on GitUpKit/Views/GIAdvancedCommitViewController.m:
```diff
@@ -212,16 +220,23 @@ - (void)_stageSelectedFiles:(NSArray*)selectedDeltas {
+        [nonSubmoduleDeltas addObject:delta];
       }
-      [deltas addObject:delta];
     }
```
Same as above.

lucasderraugh commented on GitUpKit/Core/GCIndex.m:
```diff
@@ -485,11 +485,21 @@ - (BOOL)resetLinesInFile:(NSString*)path index:(GCIndex*)index toCommit:(GCCommi
-  const char* filePath = GCGitPathFromFileSystemPath(path);
-  options.paths.strings = (char**)&filePath;
+  options.paths.count = paths.count;
+  options.paths.strings = (char**)malloc((paths.count + 1) * sizeof(char*));
```
I believe this would actually result in a memory leak. It was previously accessing the inner pointer of the string representation, no memory was being allocated and I don't believe it would be releasing it either. I also don't think there needs to be a `+ 1` here as you're giving the count information of how many `char *`s to expect so it won't be reading past the count.

lapfelix replied:
Oooh good catch. Not sure where that +1 comes from either. I removed it and added a `free` call after we call the libgit2 function. And I verified that it was being freed properly.

lapfelix commented:
I've pushed the requested change and fixed a bug in the code that calls `checkoutFilesToWorkingDirectory` (where it never sent the multiple selected file delta paths) and added comments to make it easier to understand for the next person reading it.

lucasderraugh commented:
The speed boost is fantastic. Thanks for this!

lucasderraugh commented:
@lapfelix I ended up reverting this change unfortunately. There are some major issues with discarding selected files that need resolution.

lapfelix commented:
Oh no! Do you have any steps to reproduce the issue?

lapfelix commented:
Okay I see the linked issue

lucasderraugh commented:
Ya, my repro steps were just selecting a few files and hitting the delete key to discard. I had cases where it would not notify about the change, but I also noticed that sometimes it would discard things I didn't select, so it had to be reverted for that reason.

lucasderraugh commented on GitUpKit/Views/GIAdvancedCommitViewController.m:
```diff
@@ -255,18 +264,31 @@ - (void)_diffFilesViewControllerDidPressDelete:(GIDiffFilesViewController*)contr
+
+                                       NSError* error;
+                                       if (![self discardAllChangesForFiles:selectedFiles resetIndex:NO error:&error]) {
+                                         [self.repository notifyWorkingDirectoryChanged];
```
I believe this was the issue around why it didn't notify when discarding files. This code would only notify on failures basically. The other issue where it would discard things I didn't select was the most concerning however.

lapfelix replied:
Ugh yeah the `notifyWorkingDirectoryChanged` and `makeFirstResponder` lines definitely shouldn't be in that `if`.

I'll review the rest of the code and see if I can reproduce that otgher issue where more things were discarded than necessary.

lapfelix replied:
I really can't reproduce the other issue (with these lines placed at the right place). I wonder if it might be related to your other issue where the wrong files were discarded though. I'll revert my fix and see if I can make it happen like this.

lapfelix replied:
Okaaaaayyyy I've reproduced it. It's bad.

lapfelix replied:
Looks like the issue is in the call to the libgit2 function `git_checkout_index`. Right before it's called, the repo is fine (the right files have been deleted) but it seems like it resets too many files.

lucasderraugh commented:
Ya, perhaps it’s a bug on their end? We’ll have to update libgit at some point.

lapfelix commented:
@lucasderraugh Hmmmm I have a fix which is to remove the GIT_CHECKOUT_DISABLE_PATHSPEC_MATCH option in `checkoutFilesToWorkingDirectory:` but I'll have to figure out why it was there in the first place

lapfelix commented:
@lucasderraugh I've opened a PR with the changes I'm testing: https://github.com/git-up/GitUp/pull/639

lucasderraugh commented:
Definitely want to know what that flag does. Thanks for looking into it though!

lapfelix commented:
One thing we could do to work around this apparent (?) bug is remove the flag and escape the paths we pass in so file names aren't recognized as match patterns. I've tried it and it works. I'll try it with filenames that might get recognized as match patterns before committing this change.

lucasderraugh commented:
I'll have to look up what this match pattern stuff means 😊. One other concern I have that I'd like to test out is what happens on failure? Previously just 1 file was skipped over, but in this case does every file fail to get merged? I'm thinking of the discard case where perhaps you have submodules and you try to discard them and a few other files but one thing fails so the entire discard fails.

lapfelix commented:
@lucasderraugh Submodules are still handled individually so if there's an error discarding a submodule, it will happen before we batch-process the files so we keep the same behaviour in this specific file. It will abort once it hits an error discarding a submodule.

When discarding changes in multiple files (deleting them then doing a checkout), if there's an error when deleting the file (not sure how often this happens, but it could), it will stop deleting files and will abort, present an error and stay where it was (which means the files that were already successfully deleted will stay deleted). So in that case, we won't do the checkout for these already deleted files.

If it successfully deletes the files and then fails when doing the batch-checkout, I'm not sure what will happen. I guess it will fail for all of them, yes. I haven't encountered that yet.
